Fix for field with multiple aliases

dmoree opened the pull request:
# Description

This pull request adds the ability to alias fields multiple times in queries. Currently when translating a graqhql request into a cypher query, the field name is used as opposed to the alias provided. Since the default field resolver also uses the field name this went unnoticed as long as any field was aliased at most one time during a query. However, as noted in #350 and described in detail in https://github.com/neo4j/graphql/issues/350#issuecomment-885957460; if a field is aliased multiple times, the use of the field name in the projection causes the last defined aliased field to override the others.

Aliasing of fields also requires to modify the returned object from the `create` and `update` resolver by making the key the aliased field.

# Issue

- #350 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

darrellwarde commented on packages/graphql/tests/integration/issues/350.int.test.ts:
```diff
@@ -0,0 +1,149 @@
+        driver = await neo4j();
+        const session = driver.session();
+        try {
+            await session.run(`
```
We haven't done this elsewhere... But honestly, I'm not against it, our integration tests leave the database a real mess.

danstarns replied:
I'm not against it either, something for another time but I wonder if we can add this as a global post hook that runs once after all the tests are complete.

dmoree replied:
I originally added this in a `beforeEach` but moved it to `beforeAll` since I only had one test. I wanted this to be able to use the `toHaveLength` method to ensure I was getting back the correct number of records. In order to ensure this I had to clear the database.

It would be nice to have a utility function that clears the database if needed. Something like:
```typescript
const clearDatabase = async (driver: Driver) => {
  const session = driver.session()
  try {
    await session.run(`
      MATCH (n)
      DETACH DELETE n
    `)
  } finally {
    await session.close()
  }
}
```

You could then call this in a `beforeEach` hook when writing tests if needed and in a global `afterAll` hook to clean up the database when done.

darrellwarde replied:
So, I just checked out this branch and I just cannot get the tests (in other suites) to pass. When I remove this clear-down, all of the tests pass except for this one. 

I know it's not right and our integration tests shouldn't be so co-dependent, but for the sake of getting this merged, can you change this test so that it can work without clearing the database down? 

For example, you could give the types unique names such as "Post350" and "Comment350", and then do `MATCH (p:Post350) DETACH DELETE (p)` (and the same for comments) before the tests, instead of clearing the whole database.

dmoree replied:
Sorry; fixed in https://github.com/neo4j/graphql/pull/359/commits/a8ca8f7b4cfb7404ecde4cd0838560f4056013c5. I was having trouble with `toHaveLength`. I ultimately abandoned it, but it would be nice to have cleaner tests.

darrellwarde reviewed (changes requested):
Thank for for this @dmoree, and for the code comments explaining each step - this PR should save us a lot of time.

I've just tested this out, however, and it's still missing some cases, namely for `DateTime` and `Point` fields (and maybe more). For example, for the simple schema:

```graphql
type Movie {
    datetime: DateTime
    point: Point
}
```

If I run the query:

```graphql
{
  movies {
    d1: datetime
    d2: datetime
    p1: point {
      latitude
      longitude
    }
    p2: point {
      latitude
      longitude
    }
  }
}
```

That produces the Cypher:

```cypher
MATCH (this:Movie)
RETURN this { d1: apoc.date.convertFormat(toString(this.d1), "iso_zoned_date_time", "iso_offset_date_time"), d2: apoc.date.convertFormat(toString(this.d2), "iso_zoned_date_time", "iso_offset_date_time"), p1: { point: this.p1 }, p2: { point: this.p2 } } as this
```

Note `this.d1`, `this.p1`, etc.

This should be:

```cypher
MATCH (this:Movie)
RETURN this { d1: apoc.date.convertFormat(toString(this.datetime), "iso_zoned_date_time", "iso_offset_date_time"), d2: apoc.date.convertFormat(toString(this.datetime), "iso_zoned_date_time", "iso_offset_date_time"), p1: { point: this.point }, p2: { point: this.point } } as this
```

dmoree commented on packages/graphql/src/schema/resolvers/create.ts:
```diff
@@ -19,13 +19,14 @@
 
 export default function createResolver({ node }: { node: Node }) {
-    async function resolve(_root: any, _args: any, _context: unknown) {
+    async function resolve(_root: any, _args: any, _context: unknown, _info: GraphQLResolveInfo) {
```
Of course, I'm not sure what I was thinking here.

dmoree commented:
I had only written tests specifically for the issue raised, so there was an oversight with regards to a `datetime` and `point` field; sorry for that. It should use `field.name` anytime `varName` is being accessed. Fixed in https://github.com/neo4j/graphql/pull/359/commits/3ea67c768f8e1945a30b08033cbb2987cb7406f4.

I also added the basic tests provided above for `datetime` and `point` fields.

dmoree commented on packages/graphql/src/translate/create-projection-and-params.ts:
```diff
@@ -422,7 +415,11 @@ function createProjectionAndParams({
                     : `${key}: apoc.date.convertFormat(toString(${varName}.${key}), "iso_zoned_date_time", "iso_offset_date_time")`
```
You are correct. Anytime `varName` is accessed it should use `field.name`. Fixed in https://github.com/neo4j/graphql/commit/3ea67c768f8e1945a30b08033cbb2987cb7406f4.

darrellwarde commented on packages/graphql/src/translate/create-projection-and-params.ts:
```diff
@@ -403,26 +396,30 @@ function createProjectionAndParams({
                 dateTimeField.typeMeta.array
-                    ? `${key}: [ dt in ${varName}.${key} | apoc.date.convertFormat(toString(dt), "iso_zoned_date_time", "iso_offset_date_time") ]`
-                    : `${key}: apoc.date.convertFormat(toString(${varName}.${key}), "iso_zoned_date_time", "iso_offset_date_time")`
+                    ? `${key}: [ dt in ${varName}.${field.name} | apoc.date.convertFormat(toString(dt), "iso_zoned_date_time", "iso_offset_date_time") ]`
```
I'm assuming `key` is equivalent to `field.alias`, hence why they're interchangeable?

dmoree replied:
That is correct. The key is whatever the field that is being requested is called. So if it has an alias then the key is equivalent to the alias. The projection produced a valid cypher request before due to these [lines](https://github.com/neo4j/graphql/blob/master/packages/graphql/src/translate/create-projection-and-params.ts#L143-L148) where it checks to see if there is an alias on the field and reassigns the key to be the field name.

darrellwarde reviewed (changes requested):
Thanks for making those changes!

I'm just going to have to ask you to change the integration test slightly as per my comment. It's 100% on us to tidy up the test framework, but for now I'd appreciate it if you could adjust as suggested so we can move forward quickly.

danstarns commented on packages/graphql/tests/integration/issues/350.int.test.ts:
```diff
@@ -0,0 +1,136 @@
+
+        const query = `
+            query {
+                posts {
```
```suggestion
                posts(where: { id: "${postId}" }) {
```

danstarns reviewed (changes requested):
Hey, could you add a where to this test, it should help with the failing. Here is the failing test for your info: 

```
✕ Retain attributes when aliasing the same field multiple times in a single query (200 ms)
  ● https://github.com/neo4j/graphql/issues/350 › Retain attributes when aliasing the same field multiple times in a single query
    expect(received).toContainEqual(expected) // deep equality
    Expected value: {"content": "comment 1 content", "flagged": true}
    Received array: []
      122 |             });
      123 |             expect(result.errors).toBeFalsy();
    > 124 |             expect(result?.data?.posts[0].flaggedComments).toContainEqual({
          |                                                            ^
      125 |                 content: comment1Content,
      126 |                 flagged: true,
      127 |             });
      at packages/graphql/tests/integration/issues/350.int.test.ts:124:60
      at fulfilled (packages/graphql/tests/integration/issues/350.int.test.ts:23:58)
          at runMicrotasks (<anonymous>)
```

danstarns reviewed (approved):
Thank you <3
